src/main/java/it/pagopa/pm/gateway/service/VposDeleteService.java

```diff
@@ -0,0 +1,159 @@
+import it.pagopa.pm.gateway.utils.VPosRequestUtils;
+import it.pagopa.pm.gateway.utils.VPosResponseUtils;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.http.entity.ContentType;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+import static it.pagopa.pm.gateway.constant.Messages.*;
+import static it.pagopa.pm.gateway.constant.VposConstant.RESULT_CODE_AUTHORIZED;
+import static it.pagopa.pm.gateway.dto.enums.PaymentRequestStatusEnum.CANCELLED;
+import static it.pagopa.pm.gateway.dto.enums.RefundOutcome.KO;
+import static it.pagopa.pm.gateway.dto.enums.RefundOutcome.OK;
+
+@Service
+@Slf4j
+public class VposDeleteService {
+
+    @Value("${vpos.requestUrl}")
+    private String vposUrl;
+
+    @Autowired
+    private PaymentRequestRepository paymentRequestRepository;
+
+    @Autowired
+    private VPosRequestUtils vPosRequestUtils;
+
+    @Autowired
+    private VPosResponseUtils vPosResponseUtils;
+
+    @Autowired
+    private HttpClient httpClient;
+
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    public VposDeleteResponse startDelete(String requestId) {
+        log.info("START - Vpos refund for requestId: " + requestId);
+        PaymentRequestEntity entity = paymentRequestRepository.findByGuid(requestId);
+
+        if (Objects.isNull(entity)) {
+            log.error(REQUEST_ID_NOT_FOUND_MSG);
+            return createDeleteResponse(requestId, REQUEST_ID_NOT_FOUND_MSG, KO, null);
+        }
+
+        if (BooleanUtils.isTrue(entity.getIsRefunded())) {
+            log.info("RequestId " + requestId + " has been refunded already. Skipping refund");
+            return createDeleteResponse(requestId, String.format(TRANSACTION_ALREADY_REFUND, entity.getIdTransaction()), KO, entity);
+        }
+
+        RefundOutcome refundOutcome;
+        try {
+            StepZeroRequest stepZeroRequest = getStepZeroRequest(entity);
+            refundOutcome = executeOrderStatus(entity, stepZeroRequest);
+            if (refundOutcome.equals(OK)) {
+                refundOutcome = executeRevert(entity, stepZeroRequest);
+                if(!refundOutcome.equals(OK)) {
+                    return createDeleteResponse(requestId, entity.getErrorMessage(), refundOutcome, entity);
+                }
+            } else {
+                return createDeleteResponse(requestId, entity.getErrorMessage(), refundOutcome, entity);
+            }
+        } catch (Exception e) {
+            log.error(GENERIC_ERROR_MSG + entity.getIdTransaction() + e);
+            return createDeleteResponse(requestId, GENERIC_REFUND_ERROR_MSG + requestId, KO, entity);
+        }
+        return createDeleteResponse(requestId, null, refundOutcome, entity);
```

Comment:
```suggestion
        try {
            StepZeroRequest stepZeroRequest = getStepZeroRequest(entity);
            RefundOutcome refundOutcome = executeOrderStatus(entity, stepZeroRequest);

            if (refundOutcome.equals(OK)) {
                refundOutcome = executeRevert(entity, stepZeroRequest);
                
                // %1
                if(!refundOutcome.equals(OK)) {
                    return createDeleteResponse(requestId, entity.getErrorMessage(), refundOutcome, entity);
                } else {
                    return createDeleteResponse(requestId, null, refundOutcome, entity);
                }
                // %2
            } else {
                return createDeleteResponse(requestId, entity.getErrorMessage(), refundOutcome, entity);
            }
        } catch (Exception e) {
            log.error(GENERIC_ERROR_MSG + entity.getIdTransaction() + e);
            return createDeleteResponse(requestId, GENERIC_REFUND_ERROR_MSG + requestId, KO, entity);
        }
```

Even with this refactor I'd still suggest to move the inner branches (`%1` to `%2`) into a separate method to make the refund logic clearer

Reply:
https://github.com/pagopa/pagopa-payment-transactions-gateway/pull/199/commits/8ead6e71992c09e939b6467a60fcbc8742a05cad